draft-ietf-taps-arch.md

```diff
@@ -216,35 +216,53 @@ Transport Services implementations are meant to be flexible at connection establ
+To ensure that security protocols are not incorrectly swapped, Transport Services systems MUST only select Protocol Stacks when the transport security protocols within the stacks are identical. Transport Services systems MUST NOT automatically fall back from secure protocols to insecure protocols, or to weaker versions of secure protocols. A Transport Services system MAY allow applications to explicitly specify that fallback to a specific other version of a protocol is allowed, e.g., to allow fallback to TLS 1.2 if TLS 1.3 is not available.
 
-To preserve flexibility and compatibility with future protocols, top-level features in the Transport Services API ought to avoid referencing particular transport protocols. The mappings of these API features to specific implementations of each feature is explained in the {{?I-D.ietf-taps-impl}} along with the implications of the feature on existing protocols. It is expected that {{!I-D.ietf-taps-interface}} will be updated and supplemented as new protocols and protocol features are developed.
+## Maintain Interoperability
 
-It is important to note that neither the Transport Services API {{!I-D.ietf-taps-interface}} nor the Implementation document {{?I-D.ietf-taps-impl}} define new protocols or protocol capabilities that affect what is communicated across the network. Use of a Transport Services system does not require that a peer on the other side of a connection uses the same API or implementation. A Transport Services system acting as a connection initiator can communicate with any existing system that implements the transport protocol(s) selected by the Transport Services system. Similarly, a Transport Services system acting as a listener can receive connections for any protocol that is supported by the system from existing initiators that implement the protocol, independent of whether the initiator uses Transport Services as well or not.
+It is important to note that neither the Transport Services API {{!I-D.ietf-taps-interface}} nor the Implementation document {{?I-D.ietf-taps-impl}} define new protocols or protocol capabilities that affect what is communicated across the network. Use of a Transport Services system MUST NOT require that a peer on the other side of a connection uses the same API or implementation. A Transport Services system acting as a connection initiator can communicate with any existing system that implements the transport protocol(s) selected by the Transport Services system. Similarly, a Transport Services system acting as a listener can receive connections for any protocol that is supported by the system from existing initiators that implement the protocol, independent of whether the initiator uses a Transport Services system or not.
```

Comment:
Regarding the ability to communicate with "any existing system that implements the transport protocol(s) selected by the transport system" (or have been selected by the peer, in the passive case): this is sadly not the case, see issue #516.

We could say "With the exception of cases where applications make protocol-specific assumptions, ..."   ... but that sounds clumsy, and it also loses a benefit that we have: I think the statement IS correct for TCP. Perhaps for UDP too? Beyond that... harder. I'm not sure how to best phrase this then, sorry :(

Reply:
Or...
I think my thought was more along the lines of:
"A Transport Services system acting as a connection initiator is able to communicate with any existing system that implements the transport protocol(s) selected by the Transport Services system." 
Do I understand your comment as additionally thinking also when an endpoint asks for additional property X then if that's not supported we still don't then have a connection? 
If so... then "A Transport Services system acting as a connection initiator is able to communicate with any existing system that implements the transport protocol(s) and all required properties selected by the Transport Services system."

Reply:
Yes - as you say, "when an endpoint asks for additional property X then if that's not supported we still don't then have a connection". The concrete problem on my mind is stream numbers: if we don't expose them (and I believe we don't), a TAPS Transport System cannot talk to an SCTP or QUIC peer where the application makes assumptions about stream numbers.

Stream numbers specifically should probably be included as a protocol-specific property, to avoid disappointing people implementing QUIC, which is why I filed issue #516.

Your proposed sentence at the end here, with "...and all the required properties selected by the Transport Services system"  sounds like a good way out to me!

Reply:
I like Gorry's second sentence as well but I am confused in relation to issue #516 . Isn't the problem there rather that an application that relies on a protocol specific property can not use a TAPS system that does not expose this property? This is different from what is used by the TAPS system when setting up a connection.

Reply:
For this PR, we can then conclude that we agree that this is just an editorial fix that can be solved by using Gorry's second sentence.

@abrunstrom: regarding #516, I'm not sure I fully get what you say here because (to me) you make it sound like a local problem. It is related to setting up a connection though. I had the concrete case from NEAT in mind, where Michael Tuexen wanted stream numbers so that he could write a NEAT client that can talk to a non-NEAT SCTP server which makes assumptions about stream numbers. So, my concern is the peer relying on a protocol specific property.

Considering this example, the sentence: "A Transport Services system acting as a connection initiator can communicate with any existing system that implements the transport protocol(s) selected by the Transport Services system." is just not true unless we expose stream numbers (and whatever else a peer may expect to be used). That's how #516 and the text above are related.

Reply:
In the sense that any transport can actually define unique features, the API is never generic, however, can we just keep a general claim with respect to the architecture? What word(s) would you change?

Reply:
@gorryfair I like your second sentence and do not want to change anything. 
I just wanted to point out to @mwelzl that it does not solve his issue #516. That is a related but slightly different issue. But I do not think we need to solve that here.

Reply:
Yes, let's end this discussion and agree on @gorryfair's second sentence.

Reply:
Took second sentence here.